docs/design/2022-04-15-multi-schema-change.md

```diff
@@ -0,0 +1,153 @@
+    ALTER TABLE t ADD COLUMN c INT, ADD COLUMN d INT as (c+1);
+    -- Query OK, 0 rows affected (0.07 sec)
+    ```
+  Only adding columns and adding indexes can be cascaded, and other schema object changes cannot be cascaded. For example, modifying columns cannot be cascaded:
+    ```sql
+    ALTER TABLE t ADD COLUMN d INT, CHANGE COLUMN d e INT;
+    -- ERROR 1054 (42S22): Unknown column 'd' in 't'
+    ```
+
+- Uniqueness: Multiple changes to the same schema object in the same SQL statement are not allowed.
+    ```sql
+    CREATE TABLE t (a INT, b INT);
+    ALTER TABLE t MODIFY COLUMN a CHAR(5), DROP COLUMN a;
+    -- ERROR 1054 (42S22): Unknown column 'a' in 't'
+    ```
+
+## Proposal
+
+The implementation is based on the [online DDL architecture](https://github.com/pingcap/tidb/blob/e0c461a84cf4ad55c7b51c3f9db7f7b9ba51bb62/docs/design/2018-10-08-online-DDL.md). Similar to the DDL job, we introduce a new structure "sub-job", which represents a single schema change. As its name suggests, a job can contain zero or more sub-jobs.
+
+The Multi-Schema Change DDL jobs have the type `ActionMultiSchemaChange`. In current worker model, there is a dedicated code path (`onMultiSchemaChange()`) to run these jobs. Only Multi-Schema Change job can have sub-jobs.
+
+For example, the DDL
+
+```sql
+ALTER TABLE t ADD COLUMN b INT, MODIFY COLUMN a CHAR(255);
+```
+
+can be modeled as a job like
+
+```go
+job := &Job {
+    Type: ActionMultiSchemaChange,
+    MultiSchemaInfo: &MultiSchemaInfo {
+        SubJobs: []*SubJob {
+            &SubJob {
+                Type: ActionAddColumn,
+                Args: ...
+            },
+            &SubJob {
+                Type: ActionModifyColumn,
+                Args: ...
+            },
+        }
+    } 
+}
+```
+
+In this way, we pack multiple schema changes into one job. Just like the other jobs, it enters the DDL job queue stored in TiKV, and waits for a suitable worker to pick it.
+
+In the normal cases, the worker executes the sub-jobs one by one serially as if they are plain jobs. However, the thing gets complex in the abnormal cases.
+
+To ensure the atomicity of a Multi-Schema Change execution, we need to maintain the changing schema object states carefully. Take the above SQL as an example, if the 2nd sub-job `MODIFY COLUMN a CHAR(255)` failed for some reasons, the 1st sub-job should be able to revert its changes(rollback the added column `b`).
+
+This requirement means we cannot public the schema object when a sub-job is finished. Instead, the schema object should be **paused** in a "ready-to-public" state, which is still invisible to users. These schema object are finally public all at once when all the sub-jobs are confirmed to be success. This method is kind of like 2PC: the "commit" can only be started until all "prewrites" are done.
+
+Here is the table of states that should be paused under different schema changes. Note the "Next State" column shows the states that are visible to users.
```

Comment:
> Note the "Next State" column shows the states that are visible to users.

This is not the case for "Drop Column" and "Drop Index", I would suggest using 'the changes are noticeable to the users(not-visible to visible or vice verse'.